### ISSUE TITLE
Why are we bundling 3rd party vendor's packages?

### ISSUE THREAD

**Issue by junaruga** [kennethreitz/pipenv#323]
I saw `pipenv` is using bundled 3rd party (vendor)'s package internally.

```
$ tar tzvf pipenv-3.5.6.tar.gz | grep vendor | head
drwxr-xr-x kennethreitz/staff     0 2017-04-03 14:12 pipenv-3.5.6/pipenv/vendor/
...
drwxr-xr-x kennethreitz/staff     0 2017-04-03 14:12 pipenv-3.5.6/pipenv/vendor/pexpect/
...
```

https://github.com/kennethreitz/pipenv/blob/v3.5.6/pipenv/__init__.py#L10

Why are we managing those dependency packages instead of setting the dependency packages in `setup.py` `install_requires`?

The reason why I am asking this question is that this situation is painful to make a RPM package of `pipenv` for Linux distributions. 

There is a apt-get format package for Debian, Ubuntu and etc.
There is a RPM package for RPM base distribution such as Fedora Project, CentOS, and etc.
There is a brew package for Mac.

Below is a document for Fedora Project for that, as I am in Fedora Project.
I think that other Linux distributions or Mac must have similar rules for the packaging.

https://fedoraproject.org/wiki/Bundled_Libraries
> These are some of the reasons why it's painful for us to have bundled libraries in the distribution. 

Is it possible to split vendor's packages from `pipenv`?

Thanks.

**Comment by nateprewitt** [comment]
Hey @junaruga, this was a decision that was made by Kenneth is a newer release for a few reasons. The primary one is that Kenneth wants `pipenv` to function autonomously from dependencies, so you can run the get-pipenv.py script and have everything set to go regardless of the Python setup on your system.

This is the same procedure that's been used in Requests for many years, so I'm not sure we'll see a change.

**Comment by junaruga** [comment]
Hi @nateprewitt
Okay, it's no problem.

Seeing Fedora Project's existing `requests` package, there is a patch to use external packages instead of using internal package.
http://pkgs.fedoraproject.org/cgit/rpms/python-requests.git/tree/python-requests-remove-nested-bundling-dep.patch

So, maybe I will do this kind of hack to do packaging of RPM for `pipenv` on Fedora Project.

Thanks.

**Comment by junaruga** [comment]
@nateprewitt one more question.
When we want to upgrade the version of the dependency packages under the `pipenv/vendor/`.
How to do it?

Do we use the `get-pipenv.py` script?

**Comment by nateprewitt** [comment]
@junaruga, I don't know if we have a workflow for that built into pipenv for that. The versions we've chosen are intended to be pinned and may not necessarily reflect an exact release from the upstream in some cases.

You *should* be able to freely install any of those packages separately and they won't interact with the pipenv environment.

Installing a new version of pipenv will update all the dependencies to the required versions.

**Comment by junaruga** [comment]
@nateprewitt I mean I want to know the workflow as not `pipenv` users, but `pipenv` contributors or developers. 
I want to know how to check and upgrade the version of the packages under the `pipenv/vendor/`.
Is it checking manually?

**Comment by nateprewitt** [comment]
@junaruga Yeah, this is currently handled by Kenneth. We only introduced the bundling for pipenv in the last minor release, so the workflow hasn't been fully established. We'll likely choose the latest version of each package at time of a pipenv release and bundle that along with the release. That is unless there's some kind of version conflict.

There are some special cases such as `toml` which has a fairly slow release cycle but has accepted changes we've submitted. We'll likely be vendoring either the latest copy of master of a version of 0.9.2 with our patches inlined.
